third-party/crc32c/crc32c.c

```diff
@@ -0,0 +1,597 @@
+        const char *buf3 = end + SHORT;
+        do {
+            crc = crc32cx(crc, load_unaligned_i64(buf));
+            buf += 8;
+            crc1 = crc32cx(crc1, load_unaligned_i64(buf2));
+            buf2 += 8;
+            crc2 = crc32cx(crc2, load_unaligned_i64(buf3));
+            buf3 += 8;
+        } while (buf < end);
+        crc = crc32c_shift(crc32c_short, crc) ^ crc1;
+        crc = crc32c_shift(crc32c_short, crc) ^ crc2;
+        buf += SHORT * 2;
+        len -= SHORT * 3;
+    }
+    // The same shift table can be used to compute two SHORT blocks simultaneously
+    if (len >= SHORT * 2) {
+        uint32_t crc1 = 0;
+        const char *end = buf + SHORT;
+        const char *buf2 = end;
+        do {
+            crc = crc32cx(crc, load_unaligned_i64(buf));
+            buf += 8;
+            crc1 = crc32cx(crc1, load_unaligned_i64(buf2));
+            buf2 += 8;
+        } while (buf < end);
+        crc = crc32c_shift(crc32c_short, crc) ^ crc1;
+        buf += SHORT;
+        len -= SHORT * 2;
+    }
+
+    /* compute the crc on the remaining eight-byte units less than a SHORT*2
+       block */
+    const char *end = buf + len - 8;
+    while (buf <= end) {
+        crc = crc32cx(crc, load_unaligned_i64(buf));
+        buf += 8;
+    }
+    if (len & 4) {
+        crc = crc32cw(crc, load_unaligned_i32(buf));
+        buf += 4;
+    }
+    if (len & 2) {
+        crc = crc32ch(crc, load_unaligned_i16(buf));
+        buf += 2;
+    }
+    if (len & 1)
+        crc = crc32cb(crc, *buf);
+    /* return a post-processed crc */
+    return ~crc;
+}
+
+// HW feature detection
+#  ifdef __ARM_FEATURE_CRC32
+// The C code is compiled with CRC32 being required. Skip runtime dispatch.
+uint32_t crc32c(uint32_t crc, const uint8_t *buf, size_t len)
+{
+    return crc32c_armv8(crc, buf, len);
+}
+#  else
+#   define AArch64_crc 7
+static crc32c_func_t crc32c_dispatch(unsigned long hwcap)
+{
+    if (hwcap & (1 << AArch64_crc))
+        return crc32c_armv8;
+    return crc32c_sw;
+}
+// For ifdef detection below
+#   define crc32c_dispatch() crc32c_dispatch(getauxval(AT_HWCAP))
+#   define crc32c_dispatch_ifunc "crc32c_dispatch"
+#  endif
+# endif
+
+# ifdef crc32c_dispatch
+#  ifdef CRC32C_USE_IFUNC
+// ifunc dispatch
+uint32_t crc32c(uint32_t crc, const uint8_t *buf, size_t len)
+    __attribute__((ifunc (crc32c_dispatch_ifunc)));
+#  else
+// lazy wrapper dispatch
+static uint32_t crc32c_lazy(uint32_t crc, const uint8_t *buf, size_t len);
+static crc32c_func_t crc32c_func = crc32c_lazy;
+
+static uint32_t crc32c_lazy(uint32_t crc, uint8_t *buf, size_t len)
+{
+    crc32c_func = crc32c_dispatch();
+    return crc32c_func(crc, buf, len);
+}
+
+/* Compute a CRC-32C. Do a lazy dispatch based on hardware features */
+uint32_t crc32c(uint32_t crc, const uint8_t *buf, size_t len)
+{
+    return crc32c_func(crc, buf, len);
+}
+#  endif
+# endif
+
+/* Table-driven software version as a fall-back.  This is about 15 times slower
```

Comment:
Actually the comment suggests 15x faster for the hardware accelerated version which is similar with what I've seen on some older ARM cores.